File: s3/2512/2512.cpp
#include<cstdio>
#include<algorithm>

int main()
{
	int n;
	scanf("%d", &n);
	int *budget = new int [n];
	for(int i=0; i<n; i++)
		scanf("%d", &budget[i]);
	std::sort(budget, budget+n);
	int total;
	scanf("%d", &total);
	int answer, iter;
	for(iter = 0; iter < n; iter ++)
	{
		int avg = total/(n-iter);
		answer = budget[iter];
		if(avg>=answer)
			total -= answer;
		else
			break;
	}
	if(iter<n)
		printf("%d\n", total/(n-iter));
	else
		printf("%d\n", answer);

	return 0;
}

File: s3/2512/2512.py
import sys
n = int(sys.stdin.readline())
budget = list(map(int, sys.stdin.readline().split()))
budget.sort(reverse=True)
total = int(sys.stdin.readline())
answer = 0
while budget:
    avg = total//len(budget)
    answer = budget[-1]
    if avg >= answer:
        total -= answer
        budget.pop()
    else:
        break
print(total//len(budget) if len(budget)>0 else answer)
